src/core/stores/VDocument.ts

```diff
@@ -38,6 +38,34 @@ export class VDocument {
         this.range.start.parent.after(duplicatedParent);
         nextSiblings.forEach(sibling => duplicatedParent.append(sibling));
     }
+    /**
+     * Change the selection's paragraph formatting. If there is no selection,
+     * first select the parent of the range nodes.
+     *
+     * Examples:
+     *
+     * - `<paragraph>te◆xt</paragraph>` =>
+     *   `changeType(VNodeType.HEADING1)` =>
+     *   `<heading1>▶text◀</heading1>`
+     *
+     * - `<heading1>te▶xt</heading1><heading2>te◀xt</heading2>`=>
+     *   `changeType(VNodeType.PARAGRAPH)` =>
+     *   `<paragraph>te▶xt</paragraph><paragraph>te◀xt</paragraph>`
+     *
+     * @param type
+     */
+    formatParagraph(type: VNodeType): void {
+        if (this.range.isCollapsed()) {
+            this.range.select(this.range.start.parent, this.range.end.parent);
+        }
+        new Set(this.range.selectedLeaves.map(node => node.parent)).forEach(parent => {
```

Comment:
What is most costly, assigning the same property of the same object multiple times, or creating and handling a set to call it only once ? Actually, I'm not sure I understand what is going on here...

Reply:
Good question actually, I don't know.
As to what's going on:
- Taking each selected node
- Taking their parents
- Putting that in a set (_.uniq)
- Applying the new types on them